modin/core/dataframe/algebra/groupby.py

```diff
@@ -373,16 +413,201 @@ def try_filter_dict(agg_func, df):
             Aggregation function.
         df : pandas.DataFrame
             Serialized partition which contains available columns.
+        preserve_aggregation_order : bool, default: True
+            Whether to manually restore the order of columns for the result specified
+            by the `agg_func` keys (only makes sense when `agg_func` is a dictionary).
 
         Returns
         -------
-        Callable
+        callable(pandas.core.groupby.DataFrameGroupBy) -> pandas.DataFrame
             Aggregation function that can be safely applied to this particular partition.
         """
         if not isinstance(agg_func, dict):
             return agg_func
-        partition_dict = {k: v for k, v in agg_func.items() if k in df.columns}
-        return lambda grp: grp.agg(partition_dict)
+
+        grp_has_id_level = df.columns.names[0] == cls.ID_LEVEL_NAME
+        # The 'id' level prevents us from a lookup for the original
+        # partition's columns. So dropping the level.
+        partition_columns = frozenset(
+            df.columns.droplevel(0) if grp_has_id_level else df.columns
+        )
+
+        partition_dict = {k: v for k, v in agg_func.items() if k in partition_columns}
+        return cls._build_callable_for_dict(
+            partition_dict, preserve_aggregation_order, grp_has_id_level
+        )
+
+    @classmethod
+    def _build_callable_for_dict(
+        cls, agg_dict, preserve_aggregation_order=True, grp_has_id_level=False
+    ):
+        """
+        Build callable for an aggregation dictionary.
+
+        Parameters
+        ----------
+        agg_dict : dict
+            Aggregation dictionary.
+        preserve_aggregation_order : bool, default: True
+            Whether to manually restore the order of columns for the result specified
+            by the `agg_func` keys (only makes sense when `agg_func` is a dictionary).
+        grp_has_id_level : bool, default: False
+            Whether the frame we're grouping on has intermediate columns
+            (see ``cls.ID_LEVEL_NAME``).
+
+        Returns
+        -------
+        callable(pandas.core.groupby.DataFrameGroupBy) -> pandas.DataFrame
+        """
+        external_aggs = {}
+        native_aggs = {}
```

Comment:
Why are these named as external and native?

Reply:
"Native" are aggregations that are natively applied via pandas (`groupby_object.agg(native_aggs)`). "External" are aggregations that couldn't be applied this way and so require special treatment. Is 'external' a bad name for something non-native/non-standard?

Reply:
maybe rename `external_aggs` to `non_native_aggs` and add a comment what each means.

Reply:
Done.

Renamed `external_aggs -> custom_aggs` and added the following comment:
```python
        # We now filter aggregation functions into those that could be applied natively
        # using pandas (pandas_grp_obj.agg(**native_aggs)) and those that require
        # special treatment (custom_aggs).
        custom_aggs = {}
        native_aggs = {}
```